sound/soc/sof/debug.c

```diff
@@ -17,6 +17,234 @@
 #include "sof-priv.h"
 #include "ops.h"
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_IPC_FLOOD_TEST)
+#define MAX_IPC_FLOOD_DURATION_MS 1000
+#define MAX_IPC_FLOOD_COUNT 10000
+#define IPC_FLOOD_TEST_RESULT_LEN 512
+
+static int sof_debug_ipc_flood_test(struct snd_sof_dev *sdev,
+				    struct snd_sof_dfsentry *dfse,
+				    bool flood_duration_test,
+				    unsigned long ipc_duration_ms,
+				    unsigned long ipc_count)
+{
+	struct sof_ipc_cmd_hdr hdr;
+	struct sof_ipc_reply reply;
+	u64 min_response_time = U64_MAX;
+	ktime_t start, end, test_end;
+	u64 avg_response_time = 0;
+	u64 max_response_time = 0;
+	u64 ipc_response_time;
+	char  *output;
+	int len2, len3;
+	int len1 = 0;
+	int i = 0;
+	int size;
+	int ret;
+
+	/* configure test IPC */
+	hdr.cmd = SOF_IPC_GLB_TEST_MSG | SOF_IPC_TEST_IPC_FLOOD;
+	hdr.size = sizeof(hdr);
+
+	/* set test end time for duration flood test */
+	if (flood_duration_test)
+		test_end = ktime_get_ns() + ipc_duration_ms * NSEC_PER_MSEC;
+
+	/* send test IPC's */
+	while (1) {
+		start = ktime_get();
+		ret = sof_ipc_tx_message(sdev->ipc, hdr.cmd, &hdr, hdr.size,
+					 &reply, sizeof(reply));
+		end = ktime_get();
+
+		if (ret < 0)
+			break;
+
+		/* compute min and max response times */
+		ipc_response_time = ktime_to_ns(ktime_sub(end, start));
+		min_response_time = min(min_response_time, ipc_response_time);
+		max_response_time = max(max_response_time, ipc_response_time);
+
+		/* sum up response times */
+		avg_response_time += ipc_response_time;
+		i++;
+
+		/* test complete? */
+		if (flood_duration_test) {
+			if (ktime_to_ns(end) >= test_end)
+				break;
+		} else {
+			if (i == ipc_count)
+				break;
+		}
+	}
+
+	if (ret < 0)
+		dev_err(sdev->dev,
+			"error: ipc flood test failed at %d iterations\n", i);
+
+	/* compute average response time */
+	do_div(avg_response_time, i);
+
+	if (flood_duration_test) {
+		dev_dbg(sdev->dev, "IPC Flood test duration: %lums\n",
+			ipc_duration_ms);
+		len1 = snprintf(NULL, 0,
+				"IPC Flood test duration: %lums\n",
+				ipc_duration_ms);
+	}
+
+	/* clear previous test output */
+	memset(dfse->cache_buf, 0, IPC_FLOOD_TEST_RESULT_LEN);
+
+	dev_dbg(sdev->dev,
+		"IPC Flood count: %d, Avg response time: %lluns\n",
+		i, avg_response_time);
+	dev_dbg(sdev->dev, "Max response time: %lluns\n",
+		max_response_time);
+	dev_dbg(sdev->dev, "Min response time: %lluns\n",
+		min_response_time);
+
+	/* format test output */
+	len2 = snprintf(NULL, 0,
+			"IPC Flood count: %d\nAvg response time: %lluns\n",
+			i, avg_response_time);
+
+	len3 = snprintf(NULL, 0,
+			"Max response time: %lluns\nMin response time: %lluns\n",
+			max_response_time, min_response_time);
+
+	size = len1 + len2 + len3 + 1;
+
+	output = kzalloc(size, GFP_KERNEL);
+	if (!output)
+		return -ENOMEM;
+
+	if (flood_duration_test)
+		sprintf(output, "IPC Flood test duration: %lums\n",
+			ipc_duration_ms);
+
+	sprintf(output + len1,
```

Comment:
it's odd to do a dummy strncpy, kzalloc and a real strncpy. 
isn't this what kasprintf() is meant for?

Reply:
@plbossart kasprintf is meant for this and I had this in the previous version. Unfortunately, to make this part readable, I'd need atleast 3 kasprintf()'s which means 3 separate char * buffers and deal with the errors. Whereas in the current version with the dummy snpritnf()'s, i compute the total size and allocate only one buffer and only 1 mem alloc error to deal with.

Reply:
why not create a longer string, e.g.

````
output = kasprintf("
 			IPC Flood count: %d\nAvg response time: %lluns\n
                        Max response time: %lluns\nMin response time: %lluns\n",
 			i, avg_response_time,
 			max_response_time, min_response_time);
````

Reply:
@plbossart I could do something like this: 
```+               output1 = kasprintf(GFP_KERNEL, "IPC Flood test duration: %lums\n\
+                                                IPC Flood count: %d\n\
+                                                Avg response time: %lluns\n\
+                                                Max response time: %lluns\n\
+                                                Min response time: %lluns\n",
+                                                ipc_duration_ms, i,
+                                                avg_response_time,
+                                                max_response_time,
+                                                min_response_time);
```
But this makes the output look like :
IPC Flood count: 10
						 Avg response time: 57164ns
						 Max response time: 171319ns
						 Min response time: 41633ns

Reply:
yes splitting strings across lines isn't encouraged, but if this solves the potential for memory leaks and makes the flow more linear it's probably better, no? I don't have a better idea but the sprinftf with a NULL buffer, then an alloc looks overkilll

Yes, the format looks odd, but maybe you can actually use multiple strings instead of a continuation.

 output1 = kasprintf(GFP_KERNEL, "IPC Flood test duration: %lums\n"
                                                         "IPC Flood count: %d\n"

Reply:
> Yes, the format looks odd, but maybe you can actually use multiple strings instead of a continuation.

@plbossart OK I will do that. Thats how I had it earlier but a leaky version. Let me fix it